crates/file/src/object_url.rs

```diff
@@ -0,0 +1,83 @@
+use crate::{Blob, Sealed};
+use std::{ops::Deref, rc::Rc};
+
+use wasm_bindgen::UnwrapThrowExt;
+use web_sys::Url;
+
+struct ObjectUrlAllocation {
+    url: String,
+}
+
+/// This trait is used to overload the `Url::create_object_url` function, allowing a variety of
+/// types to be used to create a [`ObjectUrl`]. Ignore this, and use [`Blob`] or [`File`] to create one.
+///
+/// The trait is sealed: it can only be implemented by types in this
+/// crate, as this crate relies on invariants regarding the `JsValue` returned from `into_jsvalue`.
+///
+/// [`File`]: crate::File
+pub trait ObjectUrlTarget: Sealed {
+    fn create_new_object_url(&self) -> String;
+}
+
+impl ObjectUrlTarget for Blob {
+    fn create_new_object_url(&self) -> String {
+        Url::create_object_url_with_blob(self.as_ref()).unwrap_throw()
+    }
+}
+
+impl Sealed for web_sys::Blob {}
+impl ObjectUrlTarget for web_sys::Blob {
+    fn create_new_object_url(&self) -> String {
+        Url::create_object_url_with_blob(self).unwrap_throw()
+    }
+}
+
+// Note: some browsers support Url::create_object_url_with_source but this is deprecated!
+// https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL#using_object_urls_for_media_streams
+
+impl ObjectUrlAllocation {
+    fn allocate_new_object_url(target: &dyn ObjectUrlTarget) -> Self {
+        let url = target.create_new_object_url();
+        Self { url }
+    }
+}
+
+impl Drop for ObjectUrlAllocation {
+    fn drop(&mut self) {
+        web_sys::Url::revoke_object_url(&self.url).unwrap_throw();
+    }
+}
+
+/// A resource wrapper around [`URL.createObjectURL`] / [`URL.revokeObjectURL`].
+///
+/// A [`Blob`], in particular a [`File`], can be converted to a short URL representing its data with the above methods.
+/// An [`ObjectUrl`] can be cheaply cloned and shared and revokes the underlying URL when the last reference is dropped.
+///
+/// Note that multiple urls can be created for the same blob, without being guaranteed to be de-deduplicated.
+///
+/// [`URL.createObjectURL`]: https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL
+/// [`URL.revokeObjectURL`]: https://developer.mozilla.org/en-US/docs/Web/API/URL/revokeObjectURL
+/// [`File`]: crate::File
+#[derive(Clone)]
+pub struct ObjectUrl {
+    inner: Rc<ObjectUrlAllocation>,
+}
+
+impl ObjectUrl {
+    /// Create a new ObjectUrl from a [`Blob`] or [`File`].
+    ///
+    /// [`File`]: crate::File
+    pub fn new(target: &dyn ObjectUrlTarget) -> Self {
+        Self {
+            inner: Rc::new(ObjectUrlAllocation::allocate_new_object_url(target)),
+        }
+    }
```

Comment:
Why not just have `From` impls? `From<Blob>`, `From<File>`, etc should be fine. I think it's better to use Rust's traits instead of creating our own, unless we have to.

A macro may be used to make implementing traits easier. Rust's orphan rules will ensure that no more traits can be implemented, effectively sealing it

Reply:
BlobContents

Reply:
> For the most part, it's replicating `BlobContents`. The "overload" here is for the `Url::create_object_url` method.

I didn't know `BlobContents` also did it. I'd argue that should be changed as well but that's for another PR. From trait can also provide the overload and (imo) is better since it's part of std lib so users are more familiar with it and understand how that overload work. That may also help avoid looking at documentation to see how to construct it since the user can just use Intellisense with the From trait.

> A potential impl for `MediaSource` has been omitted because it's deprecated. Should it be added, too?

Unless someone complains and has a valid use-case, no, let's not add support deprecated APIs.